File: tangibles/spheros/spheros/src/communication/protocol/AuthenticationProtocol.java
package communication.protocol;

import java.io.IOException;
import java.net.Socket;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.google.gson.JsonElement;
import com.google.gson.JsonSyntaxException;

import communication.JsonTcpCommunication;
import driver.AppManagerImpl;
import driver.DriverInformation;

public class AuthenticationProtocol extends JsonTcpCommunication
{

	public AuthenticationProtocol( Socket sock ) throws IOException
	{
		super( sock );
	}

	public boolean authenticate()
	{
		DriverInformation info = AppManagerImpl.getInstance().driverInfo();
		sendCtrlMessage( info );
		info( "Information message sent, waiting for response" );

		JsonElement answer = this.read();
		if( answer != null )
		{
			try
			{
				AuthenticationResponse resp = _gson.fromJson( answer, AuthenticationResponse.class );
				return resp.msg.success;
			}
			catch( JsonSyntaxException e )
			{
				// Failure to authenticate
			}
		}

		return false;
	}

	private void info( String msg )
	{
		Logger.getLogger( this.getClass().getCanonicalName() ).log( Level.INFO, msg );
	}

	private class AuthenticationResponse
	{
		String flow;
		private Response msg;

		private class Response
		{
			boolean success;
			String msg;
		}
	}
}

File: tangibles/SiftDriver/SiftDriver/Communication/Protocols/AuthenticationProtocol.cs
using System;
using System.Net.Sockets;
using System.IO;
using System.Collections.Generic;
using Sifteo;
using SiftDriver.Communication;

namespace SiftDriver.Communication.Protocols
{
  public class AuthenticationProtocol : JsonTcpCommunication
  {
    public AuthenticationProtocol (TcpClient socket) : base(socket)
    {
    }

    public bool Authenticate(){
      DriverInformation info = AppManagerAccess.Instance.DriverInfo;
      this.SendCtrlMessage(info);
      Log.Debug("infomartion message sent, waiting for an answer");
      //then wait for an answer and return
      Dictionary<string,object> returned = this.Read();
      //let's check the content of this message:
      //just some retoric tests:
      if(!returned.ContainsKey("flow") || !returned["flow"].Equals("ctrl"))
      {
        string returnedStr = "";
        foreach(string key in returned.Keys){
          returnedStr += key+" = "+returned[key]+" ; ";
        }
        throw new MissingFieldException("returned",returnedStr);
      }//else

      if(!returned.ContainsKey("msg") || !returned["msg"].GetType().Equals(typeof(Dictionary<String,Object>))){
        throw new InvalidCastException ("cannot cast msg into a dictionnary<string,object>");
      }
      Dictionary<string,object> msg = (Dictionary<string, object>) returned["msg"];
      if(!msg.ContainsKey("success") || !msg["success"].GetType().Equals(typeof(bool)) || !msg.ContainsKey("msg")){
        string returnedStr = "";
        foreach(string key in msg.Keys){
          returnedStr += key+" = "+msg[key]+" ; ";
        }
        throw new MissingFieldException("msg",returnedStr);
      }
      return (Boolean) msg["success"];
    }
  }
}

